nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/flow/resource/hadoop/HDFSExternalResourceProvider.java

```diff
@@ -140,13 +146,16 @@ public InputStream fetchExternalResource(final ExternalResourceDescriptor descri
         final HdfsResources hdfsResources = getHdfsResources();
 
         try {
-            return hdfsResources.getUserGroupInformation().doAs((PrivilegedExceptionAction<FSDataInputStream>) () -> {
-                if (!hdfsResources.getFileSystem().exists(path)) {
-                    throw new IOException("Cannot find file in HDFS at location " + location);
-                }
-
-                return hdfsResources.getFileSystem().open(path, BUFFER_SIZE_DEFAULT);
-            });
+            final FSDataInputStream fsDataInputStream =
+                    hdfsResources.getUserGroupInformation().doAs((PrivilegedExceptionAction<FSDataInputStream>) () -> {
+                        if (!hdfsResources.getFileSystem().exists(path)) {
+                            throw new IOException("Cannot find file in HDFS at location " + location);
+                        }
+
+                        return hdfsResources.getFileSystem().open(path, BUFFER_SIZE_DEFAULT);
+                    });
+            // The acquired InputStream is used by the client and cannot be closed here. The decorator is responsible for that.
```

Comment:
This part of the code is difficult to understand especially if someone is not familiar with decorator patterns. I'd mention two things here.
- The FileSystem cannot be closed upon reading from the InputStream so they're bundled in together in the HDFSResourceInputStream.
- The HDFSResourceInputStream::close also closes the FileSystem

Reply:
I think one of the benefits for using a pattern is sharing the intention with one or two words more precisely. I tried to keep answering the "why"-s in the comment and do not focus on the "how" and I think it is a reasonable expectation to have an understanding of decorator if someone is willing to change this :)

Reply:
Maintaining code readability and understandability is crucial for the long-term maintainability of a project. While it's true that using well-known design patterns like the decorator pattern can help convey the underlying structure of the code to experienced developers, it's important to consider the diverse range of developers who might interact with the codebase. By providing clear and concise comments that explain the 'why' behind design decisions, even those less familiar with certain patterns can grasp the purpose and functionality of a particular piece of code. These comments serve as a form of documentation, helping new team members onboard quickly and aiding in debugging or modification scenarios. In complex scenarios like decorator patterns, where multiple classes are interacting to achieve a specific goal, comments can be an invaluable guide for developers who are trying to make sense of the code. In essence, comments that elucidate the rationale behind design choices enable better collaboration and reduce the risk of misinterpretation or unintended modifications.

While the design pattern might provide a concise description, it remains valuable to explicitly highlight the role of the **HDFSResourceInputStream as the decorator**. Additionally, it's worth clarifying that **it's the FileSystem which must remain open until the acquired InputStream is fully processed, rather than the InputStream itself**. The decorator takes on the responsibility of closing the FileSystem when the acquired InputStream is eventually closed. This specific closing logic is a customized implementation that stands apart from the general pattern description.

Reply:
I also think some additional documentation would not be useless.

Regarding the "why": the point is that the `FileSystem` cannot be closed here so I would modify the comment:
```suggestion
            // The acquired InputStream is used by the client and for this reason the FileSystem cannot be closed here. The decorator is responsible for that.
```

Regarding the "how": I would highlight which method behaves differently and does not simply delegates the call to the decorated object. It could be added in the javadoc of `HDFSResourceInputStream` or simply here.

I agree that the benefits of patterns is sharing the intention, however it is useful to highlight the additional behaviour in case of a decorator.